### ISSUE TITLE
Cross-origin error passed to componentDidCatch incorrectly

### ISSUE THREAD

**Issue by bvaughn** [facebook/react#10441]
tl;dr React is passing "A cross-origin error was thrown" to `componentDidCatch` when there are no cross-origin scripts.

See [this discussion thread](https://github.com/facebook/react/pull/10353#issuecomment-321528685) and [this repro case](https://github.com/tessin/tessin-mini).

I was able to confirm the behavior. A quick look at `onError` showed a null `event.error`.

**Comment by bvaughn** [comment]
cc @acdlite, @gaearon

**Comment by gaearon** [comment]
Could be something funky caused by e.g. eval?

**Comment by bvaughn** [comment]
Not sure I understand. Are you asking if the user-code is maybe using `eval`?

The repro is just accessing a prop on a null value.

**Comment by bvaughn** [comment]
Looks like the repro case is as simple as:
```jsx
import React, { Component } from "react";

export default class AltApp extends Component {
  state = {
    showComponentThatFails: false,
  };

  render() {
    return (
      <ErrorBoundary>
        {this.state.showComponentThatFails && <ComponentThatFails/>}
        {!this.state.showComponentThatFails && <button onClick={() => this.setState({showComponentThatFails: true})}>Click me</button>}
      </ErrorBoundary>
    )
  }
}

class ErrorBoundary extends Component {
  state = {
    error: null
  };

  componentDidCatch(error) {
    this.setState({error})
  }

  render() {
    if (this.state.error) {
      return this.state.error.message;
    }
    return this.props.children
  }
}

class ComponentThatFails extends Component {
  render() {
    const data = null;
    return <p>{data.foo}</p>;
  }
}
```

Clicking the button will show:
> A cross-origin error was thrown.

**Comment by bvaughn** [comment]
I think this has something to do with the bundling code in [the `tessin_mini` repo](https://github.com/tessin/tessin-mini). The same code in a CRA-created app works as expected.

**Comment by bvaughn** [comment]
Repro in a fiddle: [ms1zpmgz](https://jsfiddle.net/bvaughn/ms1zpmgz/)

**Comment by bvaughn** [comment]
I can also reproduce the unexpected cross-origin behavior if loading my repro via file:// but not when loaded via http://

**Comment by bvaughn** [comment]
Creating a small, standalone project with no SSR/Webpack/etc, I am able to reproduce the unexpected behavior _only_ for Chrome when loading content via `file://...`.

Here's what I'm seeing:

| Browser | standalone `file://` | standalone `http://` | [`tessin_mini`](https://github.com/tessin/tessin-mini) |
:---|:---:|:---:|:---:
Chrome  | × | ✓ | ×
Firefox  | ✓ | ✓ |  ✓ 
Safari  | ✓ | ✓ | ✓

**Comment by gaearon** [comment]
By eval I meant the Webpack `devtool: "eval"` setting which is commonly used and turns every module into eval call so that it's shown separately in DevTools despite no sourcemaps. But seems like we determined that isn't the issue.

**Comment by bvaughn** [comment]
😭  Thanks for clarifying!

You're absolutely right. That _was_ the cause.

**Comment by gaearon** [comment]
We can probably encourage people to stop using it and use `cheap-module-source-map` instead. It's just as fast (afaik) and doesn't have some issues.

**Comment by bvaughn** [comment]
It's interesting that I use the `devtool: "eval"` setting in react-virtualized's Webpack config also but it doesn't show this error.

I'm not super familiar with Webpack. 😄

**Comment by bvaughn** [comment]
For what it's worth, I ran through the all of [the `devtools` settings](https://webpack.js.org/configuration/devtool/#devtool) and here's what I found:

devtool | works?
---|---
_default_ | ✖
eval | ✖
cheap-eval-source-map | ✖
cheap-source-map | ✓
cheap-module-eval-source-map | ✖
cheap-module-source-map | ✓
eval-source-map | ✖
source-map | ✓
inline-source-map | ✓
hidden-source-map | ✓
nosources-source-map | ✓

**Comment by bvaughn** [comment]
I'll put up a PR that adds addition verbiage to the cross-origin (dev-mode) error to mention this.

**Comment by bvaughn** [comment]
FYI @leidegre the solution for this issue is to replace:
```js
devtool: "eval"
```
with:
```js
devtool: "cheap-module-source-map"
```
in your Webpack config.

**Comment by bvaughn** [comment]
The team chatted out of band about this briefly and the agreed-upon solution is to change the wording of the error message passed to `componentDidCatch` to be:

> A cross-origin error was thrown so React doesn't have access to the actual error object in development. See https://fb.me/react-crossorigin-error for more details.

At the specified URL we'll have a blurb that explains the technique we're using in dev-mode and mentions both the `<script>` tag `crossorigin` attribute for CDNs and the Webpack `devtools` setting.

**Comment by leidegre** [comment]
@bvaughn So I'm just super happy that we could root cause this, thanks for the hard work!

**Comment by bvaughn** [comment]
You're welcome @leidegre!

Just to close the loop here, we've added a new docs page with information about this and other cross-origin error causes:

https://fb.me/react-crossorigin-error

**Comment by leidegre** [comment]
LGTM!

**Comment by sultan99** [comment]
The problem `A cross-origin error was thrown. React doesn't have access to the actual error object in development.` comes with react v16 in development mode as soon as I switch it to the production mode it works fine.

No one of these helped:
 - 'Access-Control-Allow-Origin': '*'
 - <script crossorigin="anonymous" ...
 - devtool: 'cheap-module-source-map'

**Comment by jjjjw** [comment]
I ran into this problem when using webpack, code splitting, and webpack dev server (similar to a CDN setup).

In my setup the initial bundle is loaded with a script tag, and other bundles are then loaded via JSONp by webpack. With the `crossorigin` attribute added to the script tag for the initial bundle errors can be handled, but only if they are thrown by code in the initial bundle.

To handle errors in the other bundles it's necessary to configure webpack for crossorigin script loading with this option https://webpack.js.org/configuration/output/#output-crossoriginloading

Like so in the webpack config:

```
  ...  
  output: {
    crossOriginLoading: 'anonymous',
    ...
  }
```

This adds the `crossorigin` attribute to the JSONp script tags used to load the bundles.

**Comment by bvaughn** [comment]
Hey @jjjjw. Thanks for the additional info!

Any chance you could point me to a minimal repro/setup of the setup you're describing? I'd like to better understand it so that I can update the cross-origin-errors page in the React docs.

**Comment by jjjjw** [comment]
@bvaughn 

Sure, here is a simple demo project: https://github.com/jjjjw/crossorigin-webpack-demo

Worth noting that the case is when webpack dev server is used on a different port/domain than the web server.

Additionally, here is a first go at a doc addition: https://github.com/reactjs/reactjs.org/pull/187

**Comment by amertak** [comment]
I am experiencing the same issue. Webpack is running on different domain than the whole application. It is being wired together using the nginx.
The only way to see the error for me is to start chrome with --disable-web-security flag 
I actually have no idea if its even possible to fix it.

**Comment by thomasdavis** [comment]
I'm also having no luck, using `crossorigin` attr, CORS enabled and using `cheap-module-source-maps`

**Comment by bvaughn** [comment]
Can you share a repro, @thomasdavis?

**Comment by jquense** [comment]
FYI I don't think i reopening this intentionally 😳

**Comment by bvaughn** [comment]
No worries. If people are still seeing it, it's probably okay to re-open it. 😄

**Comment by jozefpetro** [comment]
Same problem here, tried both crossOriginLoading: 'anonymous' and devtool set to 'cheap-module-eval-source-map'

**Comment by nick** [comment]
I had to do these 3 things together:

Add this to webpack config:

    devServer: {
        headers: {
            'Access-Control-Allow-Origin': '*'
        }
    }

use `cheap-module-source-map`

Add `crossorigin` to my script tag: 

    <script crossorigin src="http://localhost:8080/public/bundle.js"></script>

**Comment by loconluis** [comment]
I am using ```cheap-module-source-map```, but the error is thrown twice. Does anyone know what it could be?

**Comment by bvaughn** [comment]
> I am using `cheap-module-source-map`, but the error is thrown twice. Does anyone know what it could be?

What makes you say the error is thrown twice?

Are you referring to what Dan describes in facebook/react/issues/10384, by chance?

**Comment by herodrigues** [comment]
Same happening here using `cheap-module-source-map`, using webpack and an ErrorBoundary component.
It only happens when I don't set NODE_ENV to production, which does not make any sense since I want to run in development mode.

**Comment by bvaughn** [comment]
> It only happens when I don't set NODE_ENV to production, which does not make any sense since I want to run in development mode.

Are you also referring to what Dan describes in facebook/react#10384, by chance?

**Comment by herodrigues** [comment]
@bvaughn yeah, I've read the whole discussion in your previous comment and the answer is 'kind of'.

So, basically I shouldn't set NODE_ENV to development in my React app, because component errors should be thrown only production, right?

**Comment by bvaughn** [comment]
> So, basically I shouldn't set NODE_ENV to development in my React app

You shouldn't set `NODE_ENV` to "development" for your production app because this will result in your app [getting the slower, development build of react](https://unpkg.com/react-dom@16.0.0/index.js). The development build of React includes many helpful warnings. While useful in development, they make React larger and slower so you don't want to use them in production.

> because component errors should be thrown only production, right?

Errors can occur in development or production modes. The duplicate console logging only happens in development mode though, for reasons Dan mentioned in the issue I linked to.

**Comment by herodrigues** [comment]
I want to see those helpful warnings and I didn't set `NODE_ENV` to 'development' in production mode. 
However, my app only works **if** I set to `NODE_ENV` to 'production' in my webpack configuration like this:

```js
new DefinePlugin({
  'process.env': {
     NODE_ENV: JSON.stringify('production')
   }
})
```

If I don't use that, I see that cross-origin error.

Anyway, thanks for the explanation.
I'll read those discussions a bit more, because maybe I'm misunderstading some concepts.

**Comment by herodrigues** [comment]
Thanks @bvaughn.Z

**Comment by bvaughn** [comment]
> However, my app only works if I set NODE_ENV to 'production' even when I want to use development mode (as stated in here)

If I'm understanding this correctly, I don't know why this would be. Maybe related to another dependency your project has? Either way, sounds like a separate issue from the component stack stuff 😄 Maybe file it separately, with more details?

**Comment by herodrigues** [comment]
@bvaughn that's what I'm looking for. My app is inside a Chrome extension, which is an odd environment for React.

If it's something related to React and Chrome extensions and not an specific issue with my project, I'll create a separate issue with maybe a tutorial on how to solve it.

Thanks!

**Comment by bvaughn** [comment]
> My app is inside a Chrome extension, which is an odd environment for React.

Sounds similar to [`react-devtools`](https://github.com/facebook/react-devtools) 😄

**Comment by herodrigues** [comment]
@bvaughn, my project is buggy.
I just did a working example of a minimal Chrome extension setup using React and error boundaries.

![ezgif com-crop](https://user-images.githubusercontent.com/1832537/32628181-c89f1542-c573-11e7-981e-812fab9ff859.gif)

**Comment by DominikSerafin** [comment]
Same issue here.

Webpack+React always throws `Error: A cross-origin error was thrown. React doesn't have access to the actual error object in development.`.

Tried `devtool: 'cheap-module-source-map'`

Tried 
```
devServer: {
    headers: {
      'Access-Control-Allow-Origin': '*',
      "Access-Control-Allow-Methods": "GET, POST, PUT, DELETE, PATCH, OPTIONS",
      "Access-Control-Allow-Headers": "X-Requested-With, content-type, Authorization"
    },
}
```

Tried

```
output: {
  crossOriginLoading: 'anonymous',
}
```

Tried crossorigin script tag



Nothing works

**Comment by StalkAlex** [comment]
Experiencing same problem in Chrome, but there is no such error in Firefox.

**Comment by EdonGashi** [comment]
Getting same behavior as two previous comments in chrome.

**Comment by the-spyke** [comment]
Updated dependencies today and started to get this issue in tests. Not sure what is `just as fast`, but in my project `eval` is 2-3 times faster in rebuilds by `webpack-dev-server` watch.

**Comment by DominikSerafin** [comment]
Same thing for me as for @the-spyke, eval is much faster than other alternatives in my project(s).

React should work correctly with it, and not suggest workarounds (which do not even work).

Hoping it will get fixed...

**Comment by sapegin** [comment]
I’ve noticed that I have this issue when Chrome DevTools are open, `crossOriginLoading` and `devtool` = `cheap-module-source-map` doesn’t change anything.

**Comment by AndrewMorsillo** [comment]
I'm still getting this as well no matter my webpack settings.

**Comment by IanVS** [comment]
@herodrigues were you able to get valid errors in your chrome extension in dev?  I have a chrome extension as well and am seeing the same cross-origin errors.

**Comment by BrendonSled** [comment]
Try checking your package.json, remove the -d flag.

That's what finally worked for me.

**Comment by DominikSerafin** [comment]
@BrendonSled 

Do you mean the npm loglevel flag? https://docs.npmjs.com/misc/config#shorthands-and-other-cli-niceties

If yes, then it isn't a fix, but just merely silencing the problematic error (and probably other errors & logs too).

**Comment by herodrigues** [comment]
@IanVS there was something wrong in my project. I developed a "clean" extension using React and everything is ok.

https://github.com/herodrigues/chrome-react-minimal

**Comment by BrendonSled** [comment]
@DominikSerafin no I mean for webpack cli, https://webpack.js.org/api/cli/#shortcuts

**Comment by dehghani-mehdi** [comment]
I just added `crossorigin` to `script`  and now React tell me the error. btw I'm using following in my `package.json`

    "start": "webpack-dev-server --devtool source-map --history-api-fallback --hot"

**Comment by igorushko** [comment]
Please, make an option for disable this behavior. I want to use `eval` (performance, compatibility). And if switch to a `cheap-module-source-map`, I just get a duplicate error in the console.

**Comment by VideoCasino** [comment]
Needs investigation

**Comment by Omelettesheep** [comment]
When i use DllReferencePlugin , the problem will happen. And have tried the ways above, none of them works......

**Comment by programmerwy** [comment]
> For what it's worth, I ran through the all of [the `devtools` settings](https://webpack.js.org/configuration/devtool/#devtool) and here's what I found:
> 
> devtool	works?
> _default_	✖
> eval	✖
> cheap-eval-source-map	✖
> cheap-source-map	✓
> cheap-module-eval-source-map	✖
> cheap-module-source-map	✓
> eval-source-map	✖
> source-map	✓
> inline-source-map	✓
> hidden-source-map	✓
> nosources-source-map	✓
> It looks like the _default_ setting (and any of the *eval ones) will cause problems.

i try the table above ，bug i get the inverse result 🤣, im confused..

**Comment by fgblomqvist** [comment]
Want to highlight, similarly to @programmerwy, that it started working for me, even with the source map set to `eval-source-map`, once I added `crossorigin` to the script tag that loads the webpack bundle (couldn't get it to work no matter what without that attribute on the script tag).

Using Chrome 80.0.3987.100 (Linux) and React 16.12.

**Comment by WolvenSpirit** [comment]
This error is still thrown in React 17.0.2 when calling JSON.parse() twice on a state property, I fail to see how is it related with the actual error.

**Comment by wenfangdu** [comment]
After getting stuck on this for several days, we tried all kinds of solutions but to no avail. In the end, we found out that somehow `127.0.0.1` and `localhost` are "cross-origin", e.g. serving the app via `127.0.0.1` but visiting it via `localhost` will trigger this issue, hope that helps.

**Comment by xiongwei9** [comment]
> After getting stuck on this for several days, we tried all kinds of solutions but to no avail. In the end, we found out that `127.0.0.1` and `localhost` are cross-origin, e.g. serving the app via `127.0.0.1` but visiting it via `localhost` will trigger this issue, hope that helps.

Yes, I have the same issue. I visit my devServer `xxx.com` with http-proxy, but js script is in `localhost:3000`
![image](https://github.com/facebook/react/assets/19465270/8233317c-54af-4017-97cb-ecef7b8a2db0)

fix output of webpack config:
```
output: {
  publicPath: '/'
}
```
